### PR TITLE
HTCONDOR-2453 avoid-link-local

### PR DESCRIPTION
**Pull request by JaimeFrey**
<Insert PR description here, and leave checklist below for code review.>

# HTCondor Pull Request Checklist for internal reviewers

- [x] Verify that (GitHub thinks) the merge is clean. If it isn't, and you're confident you can resolve the conflicts, do so. Otherwise, send it back to the original developer.
- [x] Verify that the related Jira ticket exists and has a target version number and that it is correct.
- [x] Verify that the Jira ticket is in review status and is assigned to the reviewer.
- [x] Verify that the Jira ticket (HTCONDOR-xxx) is mentioned at the beginning of the title. Edit it, if not
- [x] Verify that the branch destination of the PR matches the target version of the ticket
- [x] Check for correctness of change
- [x] Check for regression test(s) of new features and bugfixes (if the feature doesn't require root)
- [x] Check for documentation, if needed  (documentation [build logs](https://readthedocs.org/projects/htcondor/builds/))
- [x] Check for version history, if needed
- [x] Check BaTLab dashboard for successful build (https://batlab.chtc.wisc.edu/results/workspace.php) and test for either the PR or a workspace build by the developer that has the Jira ticket as a comment.
- [x] Check that each commit message references the Jira ticket (HTCONDOR-xxx)

## After the above
- Hit the merge button if the pull request is approved and it is not a security patch (security changes require 2 additional reviews)
- If the pull request is approved, take the ticket out of review state
- Assign JIRA Ticket back to the developer

### CODE REVIEW

File: src/condor_utils/ipv6_addrinfo.cpp [hunk 1]
```diff
@@ -300,11 +66,11 @@ double getaddrinfo_slow_limit = 2.0;
 void (*getaddrinfo_slow_callback)(const char *node, const char *service, double timediff) = NULL;
 
 int ipv6_getaddrinfo(const char *node, const char *service,
-		addrinfo_iterator& ai, const addrinfo& hint)
+		addrinfo*& ai, const addrinfo& hint)
 {
-	addrinfo* res = NULL;
+	ai = nullptr;
```
Comment (Todd-L-Miller) [inline src/condor_utils/ipv6_addrinfo.cpp:71]: Isn't usually bad form to unconditionally modify an output parameter?

File: src/condor_utils/ipv6_hostname.cpp [hunk 2]
```diff
@@ -176,20 +176,20 @@ std::string get_fqdn_from_hostname(const std::string& hostname) {
...
+		int res  = ipv6_getaddrinfo(hostname.c_str(), NULL, info);
 		if (res) {
 			dprintf(D_HOSTNAME, "ipv6_getaddrinfo() could not look up %s: %s (%d)\n", 
 				hostname.c_str(), gai_strerror(res), res);
 			return ret;
 		}
 
-		addrinfo *info = ai.next();
-		if (info != NULL && info->ai_canonname != NULL) {
+		if (info->ai_canonname != nullptr) {
```
Comment (Todd-L-Miller) [inline src/condor_utils/ipv6_hostname.cpp:187]: This is OK if ipv6_getaddrinfo() guarantees that `info` is non-NULL, but if it it's not NULL, then, unless I'm missing something, line 189 is wrong; we need to free `info` before exiting this function, which (also) means copying `ai_canonname`.

**Review by Todd-L-Miller — COMMENTED** [review]
I had some questions; this should also have a version history entry.

**Comment by Todd-L-Miller** [comment]
This seems impractical to test.

File: src/condor_utils/ipv6_addrinfo.cpp (hunk 1 above)
Comment (JaimeFrey) [reply src/condor_utils/ipv6_addrinfo.cpp:71]: Fair point.

File: src/condor_utils/ipv6_hostname.cpp (hunk 2 above)
Comment (JaimeFrey) [reply src/condor_utils/ipv6_hostname.cpp:187]: According to the map page on macOS for getaddrinfo(), the struct addrinfo* will be non-null on a successful call. ipv6_getaddrinfo() maintains that, and we return on failure just above here. I see the linux man page isn't quite that explicit. I will fix the obvious memory leak and we can discuss the non-null assumption tomorrow.

**Comment by Todd-L-Miller** [comment]
This (still) needs a version history entry, but I have no issues with the code now.  If you'd like to remove the (hopefully) superfluous gethostbyname() calls(s), feel free and just rerequest my review.

**Review by Todd-L-Miller — APPROVED** [review]
Approved pending a successful B&T run.

**Pull request merged**